File: oauth2/src/jvmTest/java/com/okta/oauth2/kmp/jvm/AuthorizationCodeFlowTest.java
/*
 * Copyright 2022-Present Okta, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.okta.oauth2.kmp.jvm;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import com.okta.authfoundation.client.TokenInfo;
import com.okta.oauth2.kmp.BrowserRedirectHandler;
import java.util.Collections;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import org.junit.Test;

/**
 * Java-language tests for {@link AuthorizationCodeFlow} CompletableFuture wrapper. Validates that
 * the API is ergonomic from actual Java code.
 */
public class AuthorizationCodeFlowTest {

  /** A {@link BrowserRedirectHandler} stub that immediately returns a fake redirect URI. */
  private static final BrowserRedirectHandler FAKE_REDIRECT_HANDLER =
      TestFlowFactory.createFakeBrowserRedirectHandler(
          "com.example.app:/callback?code=auth-code&state=test-state");

  @Test
  public void start_ReturnsCompletableFutureWithTokenInfo()
      throws ExecutionException, InterruptedException, TimeoutException {
    try (AuthorizationCodeFlow flow = TestFlowFactory.createSuccessAuthorizationCodeFlow()) {
      CompletableFuture<TokenInfo> future =
          flow.start("com.example.app:/callback", FAKE_REDIRECT_HANDLER);
      assertNotNull("Future should not be null", future);

      TokenInfo tokenInfo = future.get(5, TimeUnit.SECONDS);
      assertNotNull("TokenInfo should not be null", tokenInfo);
      assertEquals("Bearer", tokenInfo.getTokenType());
      assertEquals("test-access-token", tokenInfo.getAccessToken());
    }
  }

  @Test
  public void start_WithAllParams_UsesProvidedScope()
      throws ExecutionException, InterruptedException, TimeoutException {
    try (AuthorizationCodeFlow flow = TestFlowFactory.createSuccessAuthorizationCodeFlow()) {
      CompletableFuture<TokenInfo> future =
          flow.start(
              "com.example.app:/callback",
              FAKE_REDIRECT_HANDLER,
              Collections.emptyMap(),
              "openid profile email");
      TokenInfo tokenInfo = future.get(5, TimeUnit.SECONDS);
      assertNotNull("TokenInfo should not be null", tokenInfo);
    }
  }

  @Test
  public void start_WithDefaults_UsesDefaultScopeAndNoExtraParams()
      throws ExecutionException, InterruptedException, TimeoutException {
    try (AuthorizationCodeFlow flow = TestFlowFactory.createSuccessAuthorizationCodeFlow()) {
      // Uses @JvmOverloads defaults — only redirectUrl and browserRedirectHandler required
      CompletableFuture<TokenInfo> future =
          flow.start("com.example.app:/callback", FAKE_REDIRECT_HANDLER);
      TokenInfo tokenInfo = future.get(5, TimeUnit.SECONDS);
      assertNotNull("TokenInfo should not be null", tokenInfo);
    }
  }

  @Test
  public void start_WithError_CompletesExceptionally()
      throws TimeoutException, InterruptedException {
    try (AuthorizationCodeFlow flow = TestFlowFactory.createFailingAuthorizationCodeFlow()) {
      CompletableFuture<TokenInfo> future =
          flow.start("com.example.app:/callback", FAKE_REDIRECT_HANDLER);
      try {
        future.get(5, TimeUnit.SECONDS);
        fail("Should have thrown ExecutionException");
      } catch (ExecutionException e) {
        assertNotNull("Cause should not be null", e.getCause());
        assertTrue(
            "Should contain error message",
            e.getCause().getMessage().contains("OIDC Endpoints not available."));
      }
    }
  }

  @Test
  public void constructor_WithOAuth2Client_CreatesFlow() {
    com.okta.authfoundation.client.kmp.OAuth2Client kmpClient = TestOAuth2Client.create();
    AuthorizationCodeFlow flow = new AuthorizationCodeFlow(kmpClient);
    assertNotNull(flow);
    flow.close();
  }

  @Test
  public void close_IsIdempotent() {
    AuthorizationCodeFlow flow = TestFlowFactory.createSuccessAuthorizationCodeFlow();
    flow.close();
    flow.close(); // Should not throw
  }
}

File: oauth2/src/commonTest/kotlin/com/okta/oauth2/kmp/AuthorizationCodeFlowTest.kt
/*
 * Copyright 2022-Present Okta, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.okta.oauth2.kmp

import com.okta.authfoundation.client.TokenInfo
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class AuthorizationCodeFlowTest {
    @Test
    fun start_WhenSuccess_ReturnsContext() =
        runTest {
            val mockFlow =
                object : AuthorizationCodeFlow {
                    override suspend fun start(
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                        scope: String,
                    ): Result<AuthorizationCodeFlowContext> =
                        Result.success(
                            AuthorizationCodeFlowContext(
                                url = "https://example.okta.com/authorize?response_type=code",
                                redirectUrl = redirectUrl,
                                codeVerifier = "test-code-verifier",
                                state = "test-state",
                                nonce = "test-nonce",
                                maxAge = null
                            )
                        )

                    override suspend fun resume(
                        uri: String,
                        flowContext: AuthorizationCodeFlowContext,
                    ): Result<TokenInfo> = Result.success(FakeTokenInfo())
                }

            val result = mockFlow.start("com.example.app:/callback")

            assertTrue(result.isSuccess)
            val context = result.getOrNull()
            assertNotNull(context)
            assertEquals("com.example.app:/callback", context.redirectUrl)
            assertTrue(context.url.contains("response_type=code"))
        }

    @Test
    fun resume_WhenSuccess_ReturnsTokenInfo() =
        runTest {
            val fakeToken = FakeTokenInfo()
            val mockFlow =
                object : AuthorizationCodeFlow {
                    override suspend fun start(
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                        scope: String,
                    ): Result<AuthorizationCodeFlowContext> =
                        Result.success(
                            AuthorizationCodeFlowContext(
                                url = "https://example.okta.com/authorize",
                                redirectUrl = "com.example.app:/callback",
                                codeVerifier = "test-code-verifier",
                                state = "test-state",
                                nonce = "test-nonce",
                                maxAge = null
                            )
                        )

                    override suspend fun resume(
                        uri: String,
                        flowContext: AuthorizationCodeFlowContext,
                    ): Result<TokenInfo> = Result.success(fakeToken)
                }

            val context =
                AuthorizationCodeFlowContext(
                    url = "https://example.okta.com/authorize",
                    redirectUrl = "com.example.app:/callback",
                    codeVerifier = "test-code-verifier",
                    state = "test-state",
                    nonce = "test-nonce",
                    maxAge = null
                )
            val uri = "com.example.app:/callback?code=auth-code&state=test-state"
            val result = mockFlow.resume(uri, context)

            assertTrue(result.isSuccess)
            val tokenInfo = result.getOrNull()
            assertNotNull(tokenInfo)
            assertEquals("test-access-token", tokenInfo.accessToken)
            assertEquals("Bearer", tokenInfo.tokenType)
        }

    @Test
    fun resume_WhenStateMismatch_ThrowsResumeException() =
        runTest {
            val mockFlow =
                object : AuthorizationCodeFlow {
                    override suspend fun start(
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                        scope: String,
                    ): Result<AuthorizationCodeFlowContext> = Result.failure(IllegalStateException())

                    override suspend fun resume(
                        uri: String,
                        flowContext: AuthorizationCodeFlowContext,
                    ): Result<TokenInfo> =
                        Result.failure(
                            AuthorizationCodeFlow.ResumeException(
                                "Failed due to state mismatch.",
                                "state_mismatch"
                            )
                        )
                }

            val context =
                AuthorizationCodeFlowContext(
                    url = "https://example.okta.com/authorize",
                    redirectUrl = "com.example.app:/callback",
                    codeVerifier = "test-code-verifier",
                    state = "expected-state",
                    nonce = "test-nonce",
                    maxAge = null
                )
            val uri = "com.example.app:/callback?code=auth-code&state=wrong-state"
            val result = mockFlow.resume(uri, context)

            assertTrue(result.isFailure)
            val exception = result.exceptionOrNull()
            assertTrue(exception is AuthorizationCodeFlow.ResumeException)
            assertEquals("state_mismatch", exception.errorId)
        }

    @Test
    fun resume_WhenErrorReturned_ThrowsResumeException() =
        runTest {
            val mockFlow =
                object : AuthorizationCodeFlow {
                    override suspend fun start(
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                        scope: String,
                    ): Result<AuthorizationCodeFlowContext> = Result.failure(IllegalStateException())

                    override suspend fun resume(
                        uri: String,
                        flowContext: AuthorizationCodeFlowContext,
                    ): Result<TokenInfo> =
                        Result.failure(
                            AuthorizationCodeFlow.ResumeException("Access denied by user.", "access_denied")
                        )
                }

            val context =
                AuthorizationCodeFlowContext(
                    url = "https://example.okta.com/authorize",
                    redirectUrl = "com.example.app:/callback",
                    codeVerifier = "test-code-verifier",
                    state = "test-state",
                    nonce = "test-nonce",
                    maxAge = null
                )
            val uri = "com.example.app:/callback?error=access_denied&error_description=Access+denied"
            val result = mockFlow.resume(uri, context)

            assertTrue(result.isFailure)
            val exception = result.exceptionOrNull()
            assertTrue(exception is AuthorizationCodeFlow.ResumeException)
            assertEquals("access_denied", exception.errorId)
        }

    @Test
    fun resume_WhenRedirectSchemeMismatch_ThrowsRedirectSchemeMismatchException() =
        runTest {
            val mockFlow =
                object : AuthorizationCodeFlow {
                    override suspend fun start(
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                        scope: String,
                    ): Result<AuthorizationCodeFlowContext> = Result.failure(IllegalStateException())

                    override suspend fun resume(
                        uri: String,
                        flowContext: AuthorizationCodeFlowContext,
                    ): Result<TokenInfo> = Result.failure(AuthorizationCodeFlow.RedirectSchemeMismatchException())
                }

            val context =
                AuthorizationCodeFlowContext(
                    url = "https://example.okta.com/authorize",
                    redirectUrl = "com.example.app:/callback",
                    codeVerifier = "test-code-verifier",
                    state = "test-state",
                    nonce = "test-nonce",
                    maxAge = null
                )
            val uri = "com.different.app:/callback?code=auth-code"
            val result = mockFlow.resume(uri, context)

            assertTrue(result.isFailure)
            assertTrue(result.exceptionOrNull() is AuthorizationCodeFlow.RedirectSchemeMismatchException)
        }

    @Test
    fun start_WhenEndpointUnavailable_ReturnsFailure() =
        runTest {
            val mockFlow =
                object : AuthorizationCodeFlow {
                    override suspend fun start(
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                        scope: String,
                    ): Result<AuthorizationCodeFlowContext> = Result.failure(IllegalStateException("OIDC Endpoints not available."))

                    override suspend fun resume(
                        uri: String,
                        flowContext: AuthorizationCodeFlowContext,
                    ): Result<TokenInfo> = Result.failure(IllegalStateException())
                }

            val result = mockFlow.start("com.example.app:/callback")

            assertTrue(result.isFailure)
            assertTrue(result.exceptionOrNull()?.message?.contains("OIDC Endpoints not available.") == true)
        }
}
